enhancements/security/runtime-class-scc.md

```diff
@@ -0,0 +1,248 @@
+
+However, instead of a customized type, or using a type from the kubernetes API like [the Runtime Class structure](https://github.com/kubernetes/api/blob/03aa42fe49ac2fa37646cc09e9a84fd825b88117/node/v1beta1/types.go#L37), `AllowedRuntimeClasses` will be a slice of strings. One can think of it as a list of [Handlers](https://github.com/kubernetes/api/blob/03aa42fe49ac2fa37646cc09e9a84fd825b88117/node/v1beta1/types.go#L53) of the runtime classes that are allowed.
+
+```
+type SecurityContextConstraints struct {
+    ...
+	// AllowedRuntimeClasses is a list of names of runtime classes (also called runtime handlers) that can be requested for a pod to run under.
+	// Runtime classes that match this list may be added to the pod request at the pod author's discretion.
+	// To allow all runtime classes, '*' may be specified.
+	// +nullable
+	AllowedRuntimeClasses []string `json:"allowedRuntimeClasses" protobuf:"bytes,26,rep,name=allowedRuntimeClasses"`
+    ...
+}
+```
+Fields in this slice will be logically OR'd. A user must specify an item that only matches one field to be permitted to use that runtime class. This allows an admin to specify a set of runtime classes that a user is allowed to use.
+
+### EnsuredRuntimeClasses
+The API for `EnsuredRuntimeClasses` will be similar in implementation the implementation of [RequiredDropCapabilities](https://github.com/openshift/api/blob/0ba2c3658da6cade27a5066575cd8446eb03914c/security/v1/types.go#L53..L55). `EnsuredRuntimeClasses` will be a slice of strings of runtime classes that must be specified for a specific SCC. Where `AllowRuntimeClasses` is something the user opts into, `EnsuredRuntimeClasses` will gate a request if the runtime class does not match one of the fields.
+
+Similarly to `AllowRuntimeClasses`, `EnsuredRuntimeClasses` will also use a string to specify the set of runtime classes the request must specify:
+```
+type SecurityContextConstraints struct {
+    ...
+	// EnsuredRuntimeClasses is a list of names of runtime classes (also called runtime handlers).     // The runtime class specified in a pod request must match one of the items in this list.
+	// To ensure any runtime class, '*' may be specified.
+	// +nullable
+	EnsuredRuntimeClasses []string `json:"ensuredRuntimeClasses" protobuf:"bytes,26,rep,name=ensuredRuntimeClasses"`
+    ...
+}
+```
+Fields in this slice will also be logically OR'd. A user must specify an item that only matches one field to satisfy the requirement. This allows an admin to specify a set of runtime classes, of which the user must choose at least one.
+
+### Some Examples
+Here are some RuntimeClasses that will be useful for Openshift to support, as well as reasonining behind not allowing everyone to use them:
+- Builder: will need access to /dev/fuse on the host, and a user namespace allocated. While these features are available to all users on the host, an untrusted user could take up all available UIDs/GIDs, preventing builds from happening.
```

Comment:
please provide full names.  Are runtime classes not logically namespaced upstream?  I expected something like `builder.openshift.io`

Reply:
no they're not, they're typically alpha strings. It's really a key->value mapping where the definition is in the crio.conf table: https://github.com/cri-o/cri-o/blob/master/docs/crio.conf.5.md#crioruntimeruntimes-table

having periods in it would mean (given toml) we'd have to have an entry like:
`[crio.runtime.runtimes."builder.openshift.io"]` I think it'd be better if we kept the simple string `builder` here

Reply:
> having periods in it would mean (given toml) we'd have to have an entry like:
> `[crio.runtime.runtimes."builder.openshift.io"]` I think it'd be better if we kept the simple string `builder` here

namespace scoping (with whatever delimiter) will prevent any conflicts from developing upstream with other tools that think they can have builder.  See our past problems with SCC/PSP names that developed.